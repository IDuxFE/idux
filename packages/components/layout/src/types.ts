import type { DefineComponent, HTMLAttributes } from 'vue'
import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'
import type { BreakpointKey } from '@idux/cdk/breakpoint'

import { IxPropTypes } from '@idux/cdk/utils'

export type SiderDirection = 'left' | 'right'
export type CollapseType = 'trigger' | 'breakpoint'

export const layoutProps = {
  siderOut: IxPropTypes.bool.def(false),
}

export const layoutHeaderProps = {
  borderless: IxPropTypes.bool.def(false),
  height: IxPropTypes.number.def(80),
}

export const layoutContentProps = {}

export const layoutSiderProps = {
  borderless: IxPropTypes.bool.def(false),
  width: IxPropTypes.number.def(272),
  direction: IxPropTypes.oneOf<SiderDirection>(['left', 'right']).def('left'),
  collapsed: IxPropTypes.bool,
  breakpoint: IxPropTypes.oneOf<BreakpointKey>(['xs', 'sm', 'md', 'lg', 'xl']),
  collapsedWidth: IxPropTypes.number.def(80),
  trigger: IxPropTypes.vNode,
  // events
  'onUpdate:collapsed': IxPropTypes.emit<(collapsed: boolean) => void>(),
  onCollapse: IxPropTypes.emit<(collapsed: boolean, type: CollapseType) => void>(),
  onBreakpoint: IxPropTypes.emit<(collapsed: boolean) => void>(),
}

export const layoutFooterProps = {
  borderless: IxPropTypes.bool.def(false),
  height: IxPropTypes.number.def(80),
}

export type LayoutProps = IxInnerPropTypes<typeof layoutProps>
export type LayoutPublicProps = IxPublicPropTypes<typeof layoutProps>
export type LayoutComponent = DefineComponent<HTMLAttributes & typeof layoutProps>
export type LayoutInstance = InstanceType<DefineComponent<LayoutProps>>

export type LayoutHeaderProps = IxInnerPropTypes<typeof layoutHeaderProps>
export type LayoutHeaderPublicProps = IxPublicPropTypes<typeof layoutHeaderProps>
export type LayoutHeaderComponent = DefineComponent<HTMLAttributes & typeof layoutHeaderProps>
export type LayoutHeaderInstance = InstanceType<DefineComponent<LayoutHeaderProps>>

export type LayoutContentProps = IxInnerPropTypes<typeof layoutContentProps>
export type LayoutContentPublicProps = IxPublicPropTypes<typeof layoutContentProps>
export type LayoutContentComponent = DefineComponent<HTMLAttributes & typeof layoutContentProps>
export type LayoutContentInstance = InstanceType<DefineComponent<LayoutContentProps>>

export type LayoutSiderProps = IxInnerPropTypes<typeof layoutSiderProps>
export type LayoutSiderPublicProps = IxPublicPropTypes<typeof layoutSiderProps>
export type LayoutSiderComponent = DefineComponent<HTMLAttributes & typeof layoutSiderProps>
export type LayoutSiderInstance = InstanceType<DefineComponent<LayoutSiderProps>>

export type LayoutFooterProps = IxInnerPropTypes<typeof layoutFooterProps>
export type LayoutFooterPublicProps = IxPublicPropTypes<typeof layoutFooterProps>
export type LayoutFooterComponent = DefineComponent<HTMLAttributes & typeof layoutFooterProps>
export type LayoutFooterInstance = InstanceType<DefineComponent<LayoutFooterProps>>
