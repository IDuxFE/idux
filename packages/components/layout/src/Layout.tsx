import { computed, defineComponent } from 'vue'
import { layoutProps } from './types'

export default defineComponent({
  name: 'IxLayout',
  props: layoutProps,
  setup(props, { slots }) {
    const classes = computed(()=>{
      return {
          'ix-layout': true,
          'sider-out': props.siderOut
      }
    })
    return ()=>{
      return (
        <section class={classes.value}>
            { slots.default?.() }
        </section>
      )
    }
  }
})
