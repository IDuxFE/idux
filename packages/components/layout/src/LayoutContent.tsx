import { defineComponent } from 'vue'
import { layoutContentProps } from './types'

export default defineComponent({
  name: 'IxLayoutContent',
  props: layoutContentProps,
  setup(props, { slots }) {
    return ()=>{
      return (
        <main class="ix-layout-content">
            { slots.default?.() }
        </main>
      )
    }
  }
})
