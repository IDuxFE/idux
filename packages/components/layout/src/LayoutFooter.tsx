import { computed, defineComponent } from 'vue'
import { layoutFooterProps } from './types'

export default defineComponent({
  name: 'IxLayoutFooter',
  props: layoutFooterProps,
  setup(props, { slots }) {
    const style = computed(()=>{
      return {
          height: `${props.height}px`,
          lineHeight: `${props.height}px`,
      }
    })

    const classes = computed(()=>{
        return {
            'ix-layout-footer': true,
            'bordered': !props.borderless
        }
    })

    return ()=>{
      return (
        <footer class={classes.value} style={style.value}>
            { slots.default?.() }
        </footer>
      )
    }
  }
})
