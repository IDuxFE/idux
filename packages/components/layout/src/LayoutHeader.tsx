import { computed, defineComponent } from 'vue'
import { layoutHeaderProps } from './types'

export default defineComponent({
  name: 'IxLayoutHeader',
  props: layoutHeaderProps,
  setup(props, { slots }) {
    const style = computed(()=>{
        return {
            height: `${props.height}px`,
            lineHeight: `${props.height}px`,
        }
    })

    const classes = computed(()=>{
        return {
            'ix-layout-header': true,
            'bordered': !props.borderless
        }
    })
    return () => {
        return (
            <header class={classes.value} style={style.value}>
                {slots.default?.()}
            </header>
        )
    }
  }
})
