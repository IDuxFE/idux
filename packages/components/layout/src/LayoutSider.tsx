import { watch } from 'vue'
import type { LayoutSiderProps, CollapseType } from './types'

import { computed, defineComponent, ref } from 'vue'
import { layoutSiderProps } from './types'
import { BREAKPOINTS, useBreakpoints } from '@idux/cdk/breakpoint'
import { callEmit, hasSlot } from '@idux/cdk/utils'
import { IxIcon } from '@idux/components/icon'
import { isUndefined } from 'lodash-es'

export default defineComponent({
  name: 'IxLayoutHeader',
  props: layoutSiderProps,
  setup(props, { slots }) {
    const isDefinedBreakPoint = !isUndefined(props.breakpoint)
    const isDefinedCollapsed = !isUndefined(props.collapsed)
    const collapsible = isDefinedBreakPoint || isDefinedCollapsed

    const hasTriggerSlot =  !!props.trigger || hasSlot(slots, 'trigger')
    const { isCollapsed, handleClick} = useCollapsed(props, isDefinedBreakPoint, isDefinedCollapsed, hasTriggerSlot)

    const style = computed(() => {
      return {
        width: isCollapsed.value ? `${Math.min(props.width, props.collapsedWidth)}px` : `${props.width}px`,
      }
    })

    const classes = computed(() => {
        return {
            'ix-layout-sider': true,
            'bordered': !props.borderless,
            'right': props.direction === 'right',
            'collapsed': collapsible && isCollapsed.value,
            'collapsed-width-zero': collapsible && isCollapsed.value && props.collapsedWidth === 0
        }
    })

    return () => {
      const trigger = collapsible && (
        props.trigger ? props.trigger :
        slots.trigger ? slots.trigger() : <IxIcon class="ix-layout-sider-trigger" name="menu" onClick={handleClick} />
      )
      return (
        <aside class={classes.value} style={style.value}>
            <div class="ix-layout-sider-children">
              { slots.default?.() }
            </div>
            { trigger }
        </aside>
      )
    }
  }
})

const useCollapsed = (props: LayoutSiderProps, isDefinedBreakPoint: boolean, isDefinedCollapsed: boolean, hasTriggerSlot:boolean) => {
  let isCollapsed = ref(false)
  let handleClick = () => {}

  if(isDefinedBreakPoint || isDefinedCollapsed) {
    isCollapsed = ref(props.collapsed ?? false)

    const changeCollapsed = (collapsed: boolean, type: CollapseType) => {
      isCollapsed.value = collapsed
      callEmit(props['onUpdate:collapsed'], collapsed)
      callEmit(props.onCollapse, collapsed, type)
      type === 'breakpoint' && callEmit(props.onBreakpoint, collapsed)
    }

    if(isDefinedBreakPoint) {
        const matchBreakpointsMaxWidth = props.breakpoint && BREAKPOINTS[props.breakpoint]?.match(/\(max-width.*\)/)?.[0]
        const breakpointsState = matchBreakpointsMaxWidth && useBreakpoints(matchBreakpointsMaxWidth)

        watch(
          () => breakpointsState && breakpointsState.matches,
          value => {
            changeCollapsed(value as boolean, 'breakpoint')
          },
          {
            // When breakpoint and collapsed exist at the same time, the collapsed value takes precedence
            immediate: !isDefinedCollapsed
          }
        )
      }

      handleClick = () => {
        changeCollapsed(!isCollapsed.value, 'trigger')
      }

      if(hasTriggerSlot && isDefinedCollapsed) {
        watch(
          () => props.collapsed,
          value => {
            changeCollapsed(value!, 'trigger')
          },
        )
      }
    }

  return {
    isCollapsed,
    handleClick,
  }
}
