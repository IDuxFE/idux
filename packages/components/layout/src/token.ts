/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { type ComputedRef, type InjectionKey } from 'vue'

import { LayoutCollapseChangeType } from './types'

export interface LayoutSiderContext {
  mergedPrefixCls: ComputedRef<string>
  collapsed: ComputedRef<boolean>
  setCollapsed: (collapsed: boolean, changeType: LayoutCollapseChangeType) => void
}

export const layoutSiderToken: InjectionKey<LayoutSiderContext> = Symbol('layoutSider')
