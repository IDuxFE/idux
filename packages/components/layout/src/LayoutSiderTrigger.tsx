/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VNodeTypes } from 'vue'

import { computed, defineComponent, inject } from 'vue'

import { isString } from 'lodash-es'

import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

import { layoutSiderToken } from './token'
import { layoutSiderTriggerProps } from './types'

export default defineComponent({
  name: 'IxLayoutSiderTrigger',
  props: layoutSiderTriggerProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-layout-sider-trigger`)
    const { collapsed, setCollapsed } = inject(layoutSiderToken)!

    const icon = computed(() => {
      const [fold = 'menu-fold', unfold = 'menu-unfold'] = props.icons || []
      return collapsed.value ? fold : unfold
    })

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return {
        [prefixCls]: true,
        [`${prefixCls}-collapsed`]: collapsed.value,
      }
    })

    const handleClick = () => setCollapsed(!collapsed.value)

    return () => {
      let children: VNodeTypes
      if (slots.default) {
        children = slots.default({ collapsed: collapsed.value })
      } else {
        const iconValue = icon.value
        children = isString(iconValue) ? <IxIcon name={iconValue} /> : iconValue
      }
      return (
        <div class={classes.value} onClick={handleClick}>
          {children}
        </div>
      )
    }
  },
})
