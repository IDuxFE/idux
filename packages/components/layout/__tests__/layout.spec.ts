import { h } from 'vue'
import { mount, MountingOptions } from '@vue/test-utils'
import { renderWork } from '@tests'
import Layout from '../src/Layout'
import LayoutContent from '../src/LayoutContent'
import LayoutSider from '../src/LayoutSider'
import LayoutHeader from '../src/LayoutHeader'
import LayoutFooter from '../src/LayoutFooter'
import { LayoutProps } from '../src/types'

const defaultSlots = [
  h(LayoutHeader, null, { default: () => 'header' }),
  h(LayoutSider, null, { default: () => 'sider' }),
  h(LayoutContent, null, { default: () => 'content' }),
  h(LayoutFooter, null, { default: () => 'footer' }),
]

describe('Layout', () => {
  renderWork<LayoutProps>(Layout, {
    slots: {
      default: () => defaultSlots,
    },
  })

  describe('basic work', () => {
    const LayoutMount = (options?: MountingOptions<Partial<LayoutProps>>) => {
      const { slots, ...rest } = options || {}
      const mergedOptions = {
        slots: { default: () => defaultSlots, ...slots },
        ...rest,
      } as MountingOptions<LayoutProps>
      return mount(Layout, mergedOptions)
    }

    test('siderOut work', async () => {
      const wrapper = LayoutMount({
        props: {
          siderOut: true,
        },
      })

      expect(wrapper.classes()).toContain('sider-out')
    })

    test('Layout-(header sider footer) borderless work', async () => {
      const wrapper = LayoutMount({
        slots: {
          default: () => [
            h(LayoutHeader, { borderless: true }, { default: () => 'header' }),
            h(LayoutSider, { borderless: true }, { default: () => 'sider' }),
            h(LayoutFooter, { borderless: true }, { default: () => 'footer' }),
          ],
        },
      })

      expect(wrapper.find('.ix-layout-header').classes()).not.toContain('bordered')
      expect(wrapper.find('.ix-layout-sider').classes()).not.toContain('bordered')
      expect(wrapper.find('.ix-layout-footer').classes()).not.toContain('bordered')
    })

    test('Layout-(header footer) height work', async () => {
      const wrapper = LayoutMount({
        slots: {
          default: () => [
            h(LayoutHeader, { height: 100 }, { default: () => 'header' }),
            h(LayoutFooter, { height: 100 }, { default: () => 'footer' }),
          ],
        },
      })

      expect(wrapper.find('.ix-layout-header').attributes()['style']).toEqual('height: 100px; line-height: 100px;')
      expect(wrapper.find('.ix-layout-footer').attributes()['style']).toEqual('height: 100px; line-height: 100px;')
    })
  })

  describe('LayoutSider', () => {
    test('direction work', async () => {
      const wrapper = mount(Layout, {
        slots: {
          default: () => [h(LayoutSider, { direction: 'right' }, { default: () => 'sider' })],
        },
      })

      expect(wrapper.find('.ix-layout-sider').classes()).toContain('right')
    })

    test('width work', async () => {
      const wrapper = mount(Layout, {
        slots: {
          default: () => [h(LayoutSider, { width: 200 }, { default: () => 'sider' })],
        },
      })

      expect(wrapper.find('.ix-layout-sider').attributes()['style']).toEqual('width: 200px;')
    })

    test('collapsed work', async () => {
      const onCollapse = jest.fn()

      const wrapper = mount(Layout, {
        slots: {
          default: () => [
            h(
              LayoutSider,
              {
                collapsed: true,
                collapsedWidth: 0,
                'onUpdate:collapsed': onCollapse,
                onCollapse,
              },
              { default: () => 'sider' },
            ),
          ],
        },
      })
      const siderWrapper = wrapper.find('.ix-layout-sider')

      expect(siderWrapper.classes()).toContain('collapsed')
      expect(siderWrapper.classes()).toContain('collapsed-width-zero')
      expect(siderWrapper.attributes()['style']).toEqual('width: 0px;')
      expect(wrapper.find('.ix-layout-sider-trigger').exists()).toBeTruthy()

      await wrapper.find('.ix-layout-sider-trigger').trigger('click')
      expect(onCollapse).toBeCalledTimes(2)
      expect(siderWrapper.classes()).not.toContain('collapsed')
      expect(siderWrapper.classes()).not.toContain('collapsed-width-zero')
    })

    test('trigger props work', async () => {
      const wrapper = mount(Layout, {
        slots: {
          default: () => [
            h(
              LayoutSider,
              {
                trigger: h('div', { class: 'vnodeTrigger' }, 'triggerDiv'),
                collapsed: true,
              },
              { default: () => 'sider' },
            ),
          ],
        },
      })

      expect(wrapper.find('.vnodeTrigger').exists()).toBeTruthy()
    })

    test('trigger slot work', async () => {
      const wrapper = mount(Layout, {
        slots: {
          default: () => [
            h(
              LayoutSider,
              {
                collapsed: true,
              },
              {
                default: () => 'sider',
                trigger: () => h('div', { class: 'slotTrigger' }, 'triggerDiv'),
              },
            ),
          ],
        },
      })

      expect(wrapper.find('.slotTrigger').exists()).toBeTruthy()
    })

    test('trigger slot and props exit the same time', async () => {
      const wrapper = mount(Layout, {
        slots: {
          default: () => [
            h(
              LayoutSider,
              {
                trigger: h('div', { class: 'vnodeTrigger' }, 'triggerDiv'),
                collapsed: true,
              },
              {
                default: () => 'sider',
                trigger: () => h('div', { class: 'slotTrigger' }, 'triggerDiv'),
              },
            ),
          ],
        },
      })

      expect(wrapper.find('.vnodeTrigger').exists()).toBeTruthy()
      expect(wrapper.find('.slotTrigger').exists()).toBeFalsy()
    })
  })
})
