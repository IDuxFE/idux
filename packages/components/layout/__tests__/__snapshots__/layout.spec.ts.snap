// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout render work 1`] = `
"<section class=\\"ix-layout\\">
  <header class=\\"ix-layout-header bordered\\" style=\\"height: 80px; line-height: 80px;\\">header</header>
  <aside class=\\"ix-layout-sider bordered\\" style=\\"width: 272px;\\">
    <div class=\\"ix-layout-sider-children\\">sider</div>
    <!---->
  </aside>
  <main class=\\"ix-layout-content\\">content</main>
  <footer class=\\"ix-layout-footer bordered\\" style=\\"height: 80px; line-height: 80px;\\">footer</footer>
</section>"
`;
